### ISSUE TITLE
IllegalStateException: Queue is empty while future is not completed

### ISSUE THREAD

**Issue by tindzk** [scalatest/scalatest#910]
I'm using ScalaTest 3.0.0-RC2 and found that my `AsyncFunSuite` test cases are failing with the following Exception:

```
java.lang.IllegalStateException: Queue is empty while future is not completed, this means you're probably using a wrong ExecutionContext for your task, please double check your Future.
```

I am not sure whether the problem is on my end as the functionality works as expected in my Scala.js test application.

Please refer to [Travis CI](https://travis-ci.org/MetaStack-pl/MetaWeb/builds/137089992) for more information.

**Comment by bvenners** [comment]
Can you paste in one of your AsyncFunSuites? The problem is indeed probably which ExecutionContext is being used.

**Comment by tindzk** [comment]
Sure. Here's the test suite that fails:

```scala
package pl.metastack.metaweb

import scala.concurrent.Future
import org.scalatest.AsyncFunSuite
import pl.metastack.metaweb.diff.{Diff, NodeRef}

class DiffSpec extends AsyncFunSuite {
  test("Replace children") {
    ...
    val future = diff.render.Tree.renderView(view)
    future.map(x => assert(x == html"""<div id="child"><span id="age">42</span><span id="name">Joe</span></div>"""))
  }
}

**Comment by bvenners** [comment]
Hmm. I can't tell from that code what execution context is being used. Is there an execution context declared in the ... part of your code? Does renderView take an implicit execution context?

**Comment by tindzk** [comment]
Ah, now I understand. Changing the interface of `renderView` to take an implicit execution context fixes the issue. Would it be possible to rephrase the error message to make it more obvious?

**Comment by fajenkyne** [comment]
We have this same ScalaTest issue with RC2.
We are trying to upgrade to RC2 because we need ScalaCheck 1.13.x to fix https://github.com/rickynils/scalacheck/issues/192.
Using ScalaTest 3.0.0-M15 (reverting to ScalaCheck 1.12.4), the code is fine.

Here's the relevant portion of our code (which is in a JVM/JS cross-platform library):

```scala
import scala.concurrent.duration.FiniteDuration
import scala.concurrent.{Future, Promise}
import scala.util.Try

trait Delayer {
  def executeAfter(delay: FiniteDuration, block: => Unit): Unit
}
object Delayer {
  private val impl: Delayer = PlatformDelayer
  def executeAfter(delay: FiniteDuration)(block: => Unit): Unit = {
    impl.executeAfter(delay, block)
  }
  def after[T](delay: FiniteDuration)(block: => T): Future[T] = {
    val promise = Promise[T]()
    executeAfter(delay)(promise.complete(Try(block)))
    promise.future
  }
}
```
On Scala.js, PlatformDelayer is implemented as follows:
```scala
import scala.concurrent.duration.FiniteDuration
import scala.scalajs.js.timers

object PlatformDelayer extends Delayer {
  override def executeAfter(delay: FiniteDuration, block: => Unit): Unit = {
    timers.setTimeout(delay)(block)
  }
}
```
And the now-failing test is:
```scala
import scala.concurrent.duration._
import org.scalatest.{AsyncFreeSpec, Matchers}

class DelayerTest extends AsyncFreeSpec with Matchers {

  "after" - {
    "should produce a future that completes but not too soon" in {
      val start = System.nanoTime().nanos
      val wait = 1.seconds
      Delayer.after(wait) {
        5
      } map { v =>
        v shouldEqual 5
        (System.nanoTime().nanos - start) should be >= wait
      }
    }
  }
}
```

**Comment by bvenners** [comment]
@fajenkyne In M15 we were using the Queue execution context by default, so if you want the same behavior, you can just override the execution context in your suite and set it to Queue, which is also ExeuctionContext.Implicits.global on Scala.js:

implicit override def executionContext =
    scala.concurrent.ExecutionContext.Implicits.global

I suspect what is going on with your code is that it is truly async, because of the timer, so you'll need to use the Queue execution context.

**Comment by fajenkyne** [comment]
@bvenners Unfortunately, using the global execution context breaks the ability to use test groups. For example, this simple, contrived test breaks (which runs fine with M15):

```scala
class FooTest extends AsyncFreeSpec {

  // commenting this out allows the test to pass
  override implicit val executionContext = scala.concurrent.ExecutionContext.Implicits.global

  def endVal(x: Int): Int = Int.MaxValue - x
  def absurd(x: Int): Future[Int] = {
    Future {
      var n = 0
      for (_ <- 0 until endVal(x)) n += 1
      n
    }
  }

  "group1" - {
    "test A" in {
      absurd(1) map { n => assert(n == endVal(1)) }
    }
  }
  "group2" - {
    "test B" in {
      absurd(2) map { n => assert(n == endVal(2)) }
    }
  }
  "group3" - {
    "test C" in {
      absurd(3) map { n => assert(n == endVal(3)) }
    }
  }
}
```

It fails with
`
java.util.ConcurrentModificationException: Two threads have apparently attempted to run a suite at the same time. This has resulted in both threads attempting to concurrently change the current Informer. 
`

But if you put all tests in one group, it succeeds.

**Comment by bvenners** [comment]
Shoot, that looks like a bug. We'll investigate, but where you actually need the Queue execution context is JS, not JVM. Do you have any way of using one execution context on the JVM and another on JS? We have a way of conditionally compiling lines of code out of, or into, the JS version. What you want to do is use the Queue execution context on JS, but you can use the default execution context on the JVM.

**Comment by fajenkyne** [comment]
@bvenners yes, we have such a way. I will pursue that as a workaround. Thanks.

**Comment by bvenners** [comment]
@fajenkyne @tindzk Can you both give 3.0.0-RC3 a try and let me know in this ticket if it solves your problems? Thanks.

**Comment by tindzk** [comment]
The problem persists I'm afraid. However, taking an implicit `ExecutionContext` is a viable solution for me.

**Comment by bvenners** [comment]
@tindzk Which problem persists? Do you mean the original code sample? If so, ScalaTest's async styles are actually supposed to work that way on Scala.js. If you really want tests to do something truly multi-threaded (and/or to use multiple execution contexts), you'll need to replace the default execution context provided by ScalaTest with one of your own on Scala.js. To use the default execution context provided by ScalaTest on Scala.js, everything involved in the test needs to use that same execution context. We did try and improve the error message in your case.

Actually I realize now that the bug we fixed was the problem that @fajenkyne reported with scopes (which he or she called " test groups"). That one was a bug, and it should be fixed in 3.0.0-RC3. @fajenkyne, please let us know if you are able to try out 3.0.0-RC3 to confirm the scopes bug is fixed. - Thanks.

**Comment by fajenkyne** [comment]
@bvenners sorry for the late reply. Yes, RC3 fixes the issue with test scopes, allowing me to provide an async ExecutionContext. I can also attest that the error message when not using a consistent ExecutionContext is better as well.
I am able to revert all the workarounds I put in place.
